File: engine/src/zobrist.cpp
#include "zobrist.h"
#include "core.h"
#include "rng.h"

#include <cassert>
#include <immintrin.h>

Zobrist::Zobrist()
{
    uint64_t seed[4] = { 1, 2, 3, 4 };
    RomuQuadRandom rng(seed);
    rng.Warm();

    for (uint64_t i = 0; i < ZOBRIST_NUMBER_COUNT; i++)
        m_Randoms[i] = rng.Generate();
}

ZobristKey Zobrist::ComputeKey(const BoardState& state) const
{
    ZobristKey key = 0;
    uint64_t offset = 0;

    // Pieces
    for (uint8_t i = 0; i < Color::MAX_ENUM; i++) {
        Color::Value color = static_cast<Color::Value>(i);
        for (uint8_t j = Piece::PAWN; j < Piece::MAX_ENUM; j++) {
            Piece::Value piece = static_cast<Piece::Value>(j);
            Bitboard occupancy = state.pieces.OccupancyMask(color, piece);
            while (occupancy) {
                uint8_t index = _tzcnt_u64(occupancy);
                key ^= m_Randoms[offset + index];
                occupancy &= (occupancy - 1);
            }
            offset += 64;
        }
    }

    // Castling rights
    if (state.rights & CastlingRight::KINGSIDE_WHITE)
        key ^= m_Randoms[offset];
    offset++;

    if (state.rights & CastlingRight::QUEENSIDE_WHITE)
        key ^= m_Randoms[offset];
    offset++;

    if (state.rights & CastlingRight::KINGSIDE_BLACK)
        key ^= m_Randoms[offset];
    offset++;

    if (state.rights & CastlingRight::QUEENSIDE_BLACK)
        key ^= m_Randoms[offset];
    offset++;

    // En Passant File
    if (state.enPassantIndex != UINT8_MAX) {
        // Check if any of our pawns can en passant the opponent pawn that just double pushed
        // This means that two identical positions (except for the en passant square) will hash to 
        // the same value as long as there is no pawn to capture en passant.
        // This check only accounts for pseudo-legal en passant captures but is better than nothing.
        uint8_t file = state.enPassantIndex & 7;
        uint8_t square = 8 * (state.turn == Color::WHITE ? 4 : 3) + file;
        uint64_t adjacentMask = 0;
        if (file > 0) adjacentMask |= (1ul << (square - 1));
        if (file < 7) adjacentMask |= (1ul << (square + 1));
        if (adjacentMask & state.pieces.OccupancyMask(state.turn, Piece::PAWN))
            key ^= m_Randoms[offset + file];
    }
    offset += 8;

    // Turn to move
    if (state.turn == Color::BLACK)
        key ^= m_Randoms[offset];
    offset++;

    assert(offset == ZOBRIST_NUMBER_COUNT && "Mistake in zobrist offset computation");

    return key;
}

File: server/src/zobrist.py
from src.board_state import BoardState, Color, Piece

class RomuRandom:
    def __init__(self, w: int, x: int, y: int, z: int):
        self._w: int = w 
        self._x: int = x 
        self._y: int = y 
        self._z: int = z 

    def warm(self, iterations: int = 10):
        for _ in range(iterations):
            _: int = self.generate()

    def generate(self) -> int:
        wp: int = self._w
        xp: int = self._x
        yp: int = self._y
        zp: int = self._z

        self._w = 15241094284759029579 * zp # a-mult
        self._x = zp + self._rotl(wp, 52) # b-rotl, c-add
        self._y = yp - xp # d-sub
        self._z = yp + wp # e-add
        self._z = self._rotl(self._z, 19) # f-rotl
        return xp

    def _rotl(self, d: int, lrot: int) -> int:
        return (d << lrot) | (d >> (64 - lrot))

class ZobristTable:
    def __init__(self):
        rng: RomuRandom = RomuRandom(0, 1, 2, 3)
        rng.warm()
        self._table: list[int] = [0] * 781
        for i in range(781):
            self._table[i] = rng.generate()

    def hash(self, state: BoardState) -> int:
        key: int = 0
        offset: int = 0

        # pieces
        for color in Color:
            for piece in Piece:
                for index in range(64):
                    p, c = state.squares[index]
                    if c == color and p == piece:
                        key ^= self._table[offset + index]
                offset += 64

        # Castling rights
        if state.rights[Color.WHITE].short:
            key ^= self._table[offset]
        offset += 1

        if state.rights[Color.WHITE].long:
            key ^= self._table[offset]
        offset += 1

        if state.rights[Color.BLACK].short:
            key ^= self._table[offset]
        offset += 1

        if state.rights[Color.BLACK].long:
            key ^= self._table[offset]
        offset += 1

        # En Passant File
        enpassant: int = state.enpassant
        if enpassant != -1:
            file: int = enpassant & 7
            if state.turn is Color.WHITE:
                index = 8 * 4 + file 
            else:
                index = 8 * 3 + file

            if file > 0 and state.squares[index - 1][0] is Piece.PAWN:
                key ^= self._table[offset + file]
            elif file < 7 and state.squares[index + 1][0] is Piece.PAWN:
                key ^= self._table[offset + file]
        offset += 8

        if state.turn is Color.BLACK:
            key ^= self._table[offset]
        offset += 1

        if offset != 781:
            raise RuntimeError("Zobrist Hashing ended with bad offset")

        return key

